src/Data/StringVariants/Util.hs

```diff
@@ -23,3 +28,39 @@ usePositiveNat n a f = case someNatVal n of
     EQI -> f p
     LTI -> f p
     GTI -> a
+
+-- Logic ported from base-4.17 Data.Char.isSpace.
+-- Hopefully, GHC would add this type-level function
+type IsCharWhitespace c
+  = If
+    (CharToNat c <=? 0x376)
+    (CharToNat c == 32 || CharToNat c - 0x9 <=? 4 || CharToNat c == 0xa0)
+    (IsUnicodeWhitespaceCodePoint (CharToNat c))
```

Comment:
Where does this `If` come from? If it is the naive one then it is strict and can result in massive blowup in compilation time, esp if nested.

Reply:
Ah, yeah, the one in `Data.Type.Bool` is the naive impl. That means we're going to evaluate both the True and False branches instead of lazily evaluating them.

This can result in a potentially large blowup in compilation time.

Inlining the comparison can help with that.

```suggestion
-- Logic ported from base-4.17 Data.Char.isSpace.
-- Hopefully, GHC would add this type-level function
type IsCharWhitespace c
  = IsCharWhitespaceImpl (CharToNat c)
  
type IsCharWhitespaceImpl (c :: Nat)
  = IsCharWhitespaceImplComp c (Compare c 0x376)
  
type family IsCharWhitespaceImplComp (c :: Nat) (cmp :: Ordering) where
  IsCharWhitespaceImplComp c 'GT =
    IsUnicodeWhitespaceCodePoint c
  IsCharWhitespaceImplComp c _ = 
    (c == 32 || c - 0x9 <=? 4 || c == 0xa0)
```

This rewrite is properly lazy, which will prevent potential blowup.

Reply:
That's interesting. Do you have a link that describes laziness on the type-level?

Reply:
[This is the best I can find](https://serokell.io/blog/type-families-haskell#evaluation-order%2C-or-lack-thereof). Seems like... the rules are entirely arbitrary.

>  Even when working with finite data, reasoning about time or space complexity of algorithms implemented as type families is impossible.

Links [this GHC issue](https://gitlab.haskell.org/ghc/ghc/-/issues/18965).